namesys/publisher.go

```diff
@@ -62,22 +67,44 @@ func (p *ipnsPublisher) PublishWithEOL(ctx context.Context, k ci.PrivKey, value
 
 	_, ipnskey := IpnsKeysForID(id)
 
-	// get previous records sequence number, and add one to it
-	var seqnum uint64
-	prevrec, err := p.routing.GetValues(ctx, ipnskey, 0)
+	// get previous records sequence number
+	seqnum, err := getPreviousSeqNo(p.ds, ipnskey)
+	if err != nil {
+		return err
+	}
+
+	// increment it
+	seqnum++
+
+	return PutRecordToRouting(ctx, k, value, seqnum, eol, p.routing, id)
+}
+
+func getPreviousSeqNo(d ds.Datastore, ipnskey key.Key) (uint64, error) {
+	prevrec, err := d.Get(ipnskey.DsKey())
 	if err == nil {
+		prbytes, ok := prevrec.([]byte)
+		if !ok {
+			return 0, fmt.Errorf("unexpected type returned from datastore: %#v", prevrec)
+		}
+		dhtrec := new(dhtpb.Record)
+		err := proto.Unmarshal(prbytes, dhtrec)
+		if err != nil {
+			return 0, err
+		}
+
 		e := new(pb.IpnsEntry)
-		err := proto.Unmarshal(prevrec[0].Val, e)
+		err = proto.Unmarshal(dhtrec.GetValue(), e)
 		if err != nil {
-			return err
+			return 0, err
 		}
 
-		seqnum = e.GetSequence() + 1
+		return e.GetSequence(), nil
 	} else if err != ds.ErrNotFound {
-		return err
+		return 0, err
```

Comment:
this won't work with a node whose local repo is destroyed, but has records published. you'll end up with records at very low seqnos that won't overwrite the others. two options: 

- request the seqnos
- use the creation timestamp as a seqno (so, two timestamps, one for creation, one for publication)